### PR TITLE
[5075] Remember user-preference for save button sub-action

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/5075

### CODE REVIEW

**Review by rart — CHANGES_REQUESTED** [review]
Please check comment on the ticket.

File: static-assets/components/cstudio-forms/forms-engine.js [hunk 1]
```diff
@@ -1868,42 +1868,30 @@ var CStudioForms =
...
-                    }
-                  },
-                  {
-                    label: formatMessage(formEngineMessages.saveAndMinimize),
-                    callback: () => {
-                      saveFn(false, true, null, 'saveAndMinimize');
-                    }
-                  }
-                ];
+                const storedId = type.previewable ? 'formEditorPreviewable' : 'formEditorNotPreviewable';
```
Comment (rart) [inline static-assets/components/cstudio-forms/forms-engine.js]: Let’s remove the save & preview option all together (everywhere)

File: static-assets/components/cstudio-forms/template-editor.js [hunk 2]
```diff
@@ -680,30 +680,16 @@ CStudioAuthoring.Module.requireModule(
```
Comment (rart) [inline static-assets/components/cstudio-forms/template-editor.js]: Where is the legacy editor still being used ?

File: ui/app/src/components/MultiChoiceSaveButton/MultiChoiceSaveButton.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  save: {
+    id: 'words.save',
+    defaultMessage: 'Save'
+  },
+  saveAndClose: {
+    id: 'codeEditor.saveAndClose',
+    defaultMessage: 'Save & Close'
+  },
+  saveAndMinimize: {
+    id: 'codeEditor.saveAndMinimize',
```
Comment (rart) [inline ui/app/src/components/MultiChoiceSaveButton/MultiChoiceSaveButton.tsx]: rename these ids

File: ui/app/src/utils/state.ts [hunk 4]
```diff
@@ -207,3 +207,15 @@ export function commonDialogProps<T>(specificProps: T): EnhancedDialogState & T
     ...specificProps
   };
 }
+
+export function setStoredSaveButtonSubAction(user: string, id: string, value: string) {
```
Comment (rart) [inline ui/app/src/utils/state.ts]:
```suggestion
export function setStoredSaveButtonSubAction(user: string, id: string, value: string): void {
```

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-forms/template-editor.js (hunk 2 above)
Comment (jvega190) [reply static-assets/components/cstudio-forms/template-editor.js]: There are some parts where it's being used but seems like dead code. I'll revert the changes. Should I create a ticket to remove template-editor.js and dead usages?

Comment (rart) [reply static-assets/components/cstudio-forms/template-editor.js]: yes

File: ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx [hunk 5]
```diff
@@ -157,7 +157,15 @@ export function SiteToolsApp(props: SiteToolsAppProps) {
...
           tool ? (
             <Suspencified>
-              <Widget {...tool} extraProps={{ embedded: false, showAppsButton, onSubmittingAndOrPendingChange }} />
+              <Widget
+                {...tool}
+                extraProps={{
+                  embedded: false,
+                  showAppsButton,
+                  onSubmittingAndOrPendingChange,
+                  configuration: { isContainerDialog: hideSidebarSiteSwitcher }
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx]: 👆🏻

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 6]
```diff
@@ -30,15 +30,18 @@ import {
   contentTypeUpdated,
   emitSystemEvent
 } from '../../state/actions/system';
+import { updateWidgetDialog } from '../../state/actions/dialogs';
+import { LookupTable } from '../../models';
 
 interface ContentTypeManagementProps {
   embedded?: boolean;
   showAppsButton?: boolean;
+  configuration?: LookupTable;
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]: 👆🏻

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 7]
```diff
@@ -58,6 +61,21 @@ export function ContentTypeManagement(props: ContentTypeManagementProps) {
...
+                break;
+              case 'saveAndMinimize':
+                if (embedded) {
+                  configuration?.onMinimize?.();
+                } else if (configuration?.isContainerDialog) {
+                  // If not embedded but container is a dialog, dispatch action to minimize.
+                  dispatch(updateWidgetDialog({ isMinimized: true }));
+                }
+                break;
+            }
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]: 👆🏻

File: ui/app/src/components/WidgetDialog/utils.ts [hunk 8]
```diff
@@ -18,10 +18,12 @@ import { WidgetDescriptor } from '../Widget';
 import StandardAction from '../../models/StandardAction';
 import { EnhancedDialogProps } from '../EnhancedDialog';
 import { EnhancedDialogState, onSubmittingAndOrPendingChangeProps } from '../../hooks/useEnhancedDialogState';
+import { LookupTable } from '../../models';
 
 interface WidgetDialogBaseProps {
   title: string;
   widget: WidgetDescriptor;
+  configuration?: LookupTable;
```
Comment (rart) [inline ui/app/src/components/WidgetDialog/utils.ts]: 👆🏻

**Review by rart — CHANGES_REQUESTED** [review]
Let's discuss.

File: static-assets/components/cstudio-forms/template-editor.js (hunk 2 above)
Comment (jvega190) [reply static-assets/components/cstudio-forms/template-editor.js]: ticket created https://github.com/craftercms/craftercms/issues/5138

File: static-assets/components/cstudio-forms/forms-engine.js [hunk 9]
```diff
@@ -1868,42 +1868,21 @@ var CStudioForms =
 
             //In Context Edit, the preview button must not be shown
             var iceId = CStudioAuthoring.Utils.getQueryVariable(location.search, 'iceId');
-
+            const user = CStudioAuthoringContext.user;
```
Comment (rart) [inline static-assets/components/cstudio-forms/forms-engine.js]: I don't see this being used.

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 10]
```diff
@@ -58,6 +60,16 @@ export function ContentTypeManagement(props: ContentTypeManagementProps) {
         switch (e.data.type) {
           case 'CONTENT_TYPES_ON_SAVED': {
             dispatch(emitSystemEvent(contentTypeUpdated()));
+            switch (e.data.saveType) {
+              case 'saveAndClose':
+                if (embedded) {
+                  onClose?.();
+                }
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]: If the saveAndClose is available and an onClose function is sent, no reason not to call it independently of whether it is embedded or not.
```suggestion
                onClose?.();
```

File: ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx [hunk 11]
```diff
@@ -157,7 +160,20 @@ export function SiteToolsApp(props: SiteToolsAppProps) {
...
           tool ? (
             <Suspencified>
-              <Widget {...tool} extraProps={{ embedded: false, showAppsButton, onSubmittingAndOrPendingChange }} />
+              <Widget
+                {...tool}
+                extraProps={{
+                  embedded: false,
+                  showAppsButton,
+                  onSubmittingAndOrPendingChange,
+                  configuration: { isContainerDialog: hideSidebarSiteSwitcher },
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx]: 👆🏻???

File: ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx [hunk 12]
```diff
@@ -157,7 +160,20 @@ export function SiteToolsApp(props: SiteToolsAppProps) {
...
+                  ...(hideSidebarSiteSwitcher && {
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx]: Why is on minimize conditional to hideSidebarSiteSwitcher?

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/SiteToolsApp/UrlDrivenSiteTools.tsx [hunk 13]
```diff
@@ -70,6 +70,7 @@ export function UrlDrivenSiteTools(props: UrlDrivenSiteToolsProps) {
       openSidebar={openSidebar || !activeToolId}
       tools={tools}
       imageUrl={`${baseUrl}/static-assets/images/choose_option.svg`}
+      urlDriven={true}
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/UrlDrivenSiteTools.tsx]:
```suggestion
      urlDriven
```

File: ui/app/src/components/ToolsPanelEmbeddedAppViewButton/ToolsPanelEmbeddedAppViewButton.tsx [hunk 14]
```diff
@@ -40,6 +40,7 @@ export function ToolsPanelEmbeddedAppViewButton(props: ToolsPanelEmbeddedAppView
   } = useEnhancedDialogState();
   const title = usePossibleTranslation(props.title);
   const widgetDialogPendingChangesCloseRequest = useWithPendingChangesCloseRequest(onClose);
+  const isWidgetEmbedded = props.widget?.configuration?.embedded;
```
Comment (rart) [inline ui/app/src/components/ToolsPanelEmbeddedAppViewButton/ToolsPanelEmbeddedAppViewButton.tsx]: This component has no business reading the internals of widgets and their config.

File: ui/app/src/components/ToolsPanelEmbeddedAppViewButton/ToolsPanelEmbeddedAppViewButton.tsx [hunk 15]
```diff
@@ -56,6 +57,11 @@ export function ToolsPanelEmbeddedAppViewButton(props: ToolsPanelEmbeddedAppView
         open={open}
         onClose={onClose}
         widget={props.widget}
+        extraProps={{
+          onMinimize,
+          onMaximize,
+          ...(isWidgetEmbedded && { onClose })
```
Comment (rart) [inline ui/app/src/components/ToolsPanelEmbeddedAppViewButton/ToolsPanelEmbeddedAppViewButton.tsx]: What's the problem with simply providing the onClose?

File: ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx [hunk 16]
```diff
@@ -58,6 +60,7 @@ export interface SiteToolsAppProps {
   hideSidebarSiteSwitcher?: boolean;
   showAppsButton?: boolean;
   classes?: Partial<Record<'root', string>>;
+  urlDriven?: boolean;
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx]:
```suggestion
  // Whether the component is mounted on a dialog or it's the main app on a page (i.e. `/studio/site-tools`)
  mountMode?: 'dialog' | 'page';
```

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 17]
```diff
@@ -34,11 +34,14 @@ import {
 interface ContentTypeManagementProps {
   embedded?: boolean;
   showAppsButton?: boolean;
+  urlDriven?: boolean;
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]: use mountMode all across

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-admin/base.js [hunk 18]
```diff
@@ -350,16 +350,33 @@
...
+          buttonEl.id = action.id;
+        }
+        this.commandBarEl.appendChild(buttonEl);
+        buttonEl.onclick = action.fn;
+      } else {
+        const mountMode = CStudioAuthoring.Utils.getQueryParameterByName('mountMode');
+        const options = ['save', 'saveAndClose'];
+        if (mountMode !== 'page') {
+          options.push('saveAndMinimize');
+        }
```
Comment (rart) [inline static-assets/components/cstudio-admin/base.js]:
```suggestion
        if (mountMode === 'dialog') {
          options.push('saveAndMinimize');
        }
```

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 19]
```diff
@@ -34,11 +34,14 @@ import {
 interface ContentTypeManagementProps {
   embedded?: boolean;
   showAppsButton?: boolean;
+  mountMode?: boolean;
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]: mountMode is not a boolean

File: ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx [hunk 20]
```diff
@@ -157,7 +163,18 @@ export function SiteToolsApp(props: SiteToolsAppProps) {
...
+              <Widget
+                {...tool}
+                extraProps={{
+                  embedded: false,
+                  mountMode,
+                  showAppsButton,
+                  onSubmittingAndOrPendingChange,
+                  onMinimize: () => {
+                    dispatch(updateWidgetDialog({ isMinimized: true }));
+                  }
```
Comment (rart) [inline ui/app/src/components/SiteToolsApp/SiteToolsApp.tsx]: Could we send this in from the parent component so we don't couple SiteToolsApp to redux?

**Review by rart — CHANGES_REQUESTED** [review]

**Comment by jvega190** [comment]
@rart since this ticket is rc2, should I change the PR to be a draft?

File: ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx [hunk 21]
```diff
@@ -89,7 +100,7 @@ export function ContentTypeManagement(props: ContentTypeManagementProps) {
       )}
       {loading && <LoadingState styles={{ root: { flexGrow: 1 } }} />}
       <LegacyIFrame
-        path="/legacy-site-config?mode=embedded#tool/content-types"
+        path={`/legacy-site-config?mode=embedded${mountMode ? '&mountMode=' + mountMode : ''}#tool/content-types`}
```
Comment (rart) [inline ui/app/src/components/ContentTypesManagement/ContentTypesManagement.tsx]:
```suggestion
        path={`/legacy-site-config?mode=embedded${mountMode ? `&mountMode=${mountMode}` : ''}#tool/content-types`}
```

**Review by rart — APPROVED** [review]

**Pull request merged**